// @univerjs/rpc/lib/es/index.js
var X = Object.defineProperty;
var Y = (s, t, e) => t in s ? X(s, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : s[t] = e;
var o = (s, t, e) => Y(s, typeof t != "symbol" ? t + "" : t, e);
import { UniverInstanceType as u, ICommandService as b, IUniverInstanceService as N, ILogService as Z, RxDisposable as q, OnLifecycle as W, LifecycleStages as L, CommandType as H, Disposable as K, Plugin as V } from "@univerjs/core";
import { createIdentifier as O, Inject as g, Injector as C } from "@wendellhu/redi";
import { takeUntil as f, filter as ee, take as te } from "rxjs/operators";
import { BehaviorSubject as ne, firstValueFrom as se, Observable as $, isObservable as re, of as ie, shareReplay as A } from "rxjs";
var ce = Object.defineProperty, ae = Object.getOwnPropertyDescriptor, F = (s, t, e, n) => {
  for (var r = n > 1 ? void 0 : n ? ae(t, e) : t, i = s.length - 1, c; i >= 0; i--)
    (c = s[i]) && (r = (n ? c(t, e, r) : c(r)) || r);
  return n && r && ce(t, e, r), r;
}, m = (s, t) => (e, n) => t(e, n, s);
const j = "rpc.remote-sync.service", y = O(j);
let M = class {
  constructor(s) {
    this._commandService = s;
  }
  async syncMutation(s) {
    return this._commandService.syncExecuteCommand(s.mutationInfo.id, s.mutationInfo.params, {
      onlyLocal: !0,
      fromSync: !0
    });
  }
};
M = F([
  m(0, b)
], M);
const x = "univer.remote-instance-service", S = O(x);
let U = class {
  constructor(s, t, e) {
    this._univerInstanceService = s, this._commandService = t, this._logService = e;
  }
  whenReady() {
    return Promise.resolve(!0);
  }
  async syncMutation(s) {
    return this._applyMutation(s.mutationInfo);
  }
  async createInstance(s) {
    const { type: t, snapshot: e } = s;
    try {
      switch (t) {
        case u.UNIVER_SHEET:
          return this._univerInstanceService.createUnit(u.UNIVER_SHEET, e), !0;
        default:
          throw new Error(
            `[WebWorkerRemoteInstanceService]: cannot create replica for document type: ${t}.`
          );
      }
    } catch (n) {
      throw n instanceof Error ? n : new TypeError(`${n}`);
    }
  }
  async disposeInstance(s) {
    return this._univerInstanceService.disposeUnit(s.unitID);
  }
  _applyMutation(s) {
    const { id: t, params: e } = s;
    return this._commandService.syncExecuteCommand(t, e, {
      onlyLocal: !0,
      fromSync: !0
    });
  }
};
U = F([
  m(0, N),
  m(1, b),
  m(2, Z)
], U);
function z(s) {
  const t = s;
  return new class {
    call(e, n) {
      const r = t[e];
      if (typeof r == "function") {
        let i = r.apply(t, [n]);
        return i instanceof Promise || (i = Promise.resolve(i)), i;
      }
      throw new Error(`[RPC]: method not found for ${e}!`);
    }
    subscribe(e, n) {
      const r = t[e];
      if (typeof r == "function") {
        const i = r.apply(t, n);
        return re(i) ? i : ie(i);
      }
      throw new Error(`[RPC]: observable method not found for ${e}!`);
    }
  }();
}
function G(s) {
  return new Proxy({}, {
    get(t, e) {
      if (e !== "dispose")
        return function(...n) {
          return oe(e) ? s.subscribe(e, n[0]) : s.call(e, n[0]);
        };
    }
  });
}
function oe(s) {
  return s.endsWith("$");
}
class he extends q {
  constructor(e) {
    super();
    o(this, "_initialized", new ne(!1));
    o(this, "_lastRequestCounter", 0);
    o(this, "_pendingRequests", /* @__PURE__ */ new Map());
    o(this, "_pendingSubscriptions", /* @__PURE__ */ new Map());
    this._protocol = e, this._protocol.onMessage.pipe(f(this.dispose$)).subscribe((n) => this._onMessage(n));
  }
  getChannel(e) {
    const n = this;
    return {
      call(r, i) {
        return n._disposed ? Promise.reject() : n._remoteCall(e, r, i);
      },
      subscribe(r, i) {
        if (n._disposed)
          throw new Error("[ChannelClient]: client is disposed!");
        return n._remoteSubscribe(e, r, i);
      }
    };
  }
  _whenReady() {
    return se(
      this._initialized.pipe(
        ee((e) => e),
        te(1)
      )
    );
  }
  async _remoteCall(e, n, r) {
    await this._whenReady();
    const i = ++this._lastRequestCounter, h = { seq: i, type: 100, channelName: e, method: n, args: r }, a = this;
    return new Promise((I, _) => {
      const Q = {
        handle(w) {
          switch (w.type) {
            case 201:
              a._pendingRequests.delete(i), I(w.data);
              break;
            case 202:
              a._pendingRequests.delete(i), _(w.data);
              break;
            default:
              throw new Error("[ChannelClient]: unknown response type!");
          }
        }
      };
      this._pendingRequests.set(i, Q), this._sendRequest(h);
    });
  }
  _remoteSubscribe(e, n, r) {
    return new $((i) => {
      let c = -1;
      return this._whenReady().then(() => {
        c = ++this._lastRequestCounter;
        const a = { seq: c, type: 101, channelName: e, method: n, args: r }, I = {
          handle(_) {
            switch (_.type) {
              case 300:
                i.next(_.data);
                break;
              case 301:
                i.error(_.data);
                break;
              case 302:
                i.complete();
                break;
              default:
                throw new Error("[ChannelClient]: unknown response type!");
            }
          }
        };
        this._pendingRequests.set(c, I), this._sendRequest(a);
      }), () => {
        if (c === -1)
          return;
        const h = {
          type: 102,
          seq: c,
          channelName: e,
          method: n
        };
        this._sendRequest(h);
      };
    });
  }
  _sendRequest(e) {
    this._protocol.send(e);
  }
  _onMessage(e) {
    var n;
    switch (e.type) {
      case 0:
        this._initialized.next(!0);
        break;
      case 201:
      case 202:
      case 300:
      case 302:
      case 301:
        (n = this._pendingRequests.get(e.seq)) == null || n.handle(e);
        break;
    }
  }
}
class _e extends q {
  constructor(e) {
    super();
    o(this, "_channels", /* @__PURE__ */ new Map());
    o(this, "_subscriptions", /* @__PURE__ */ new Map());
    this._protocol = e, this._protocol.onMessage.pipe(f(this.dispose$)).subscribe((n) => this._onRequest(n)), this._sendResponse({
      seq: -1,
      type: 0
      /* INITIALIZE */
    });
  }
  registerChannel(e, n) {
    this._channels.set(e, n);
  }
  _onRequest(e) {
    switch (e.type) {
      case 100:
        this._onMethodCall(e);
        break;
      case 101:
        this._onSubscribe(e);
        break;
      case 102:
        this._onUnsubscribe(e);
        break;
    }
  }
  _onMethodCall(e) {
    const { channelName: n, method: r, args: i } = e, c = this._channels.get(n);
    let h;
    try {
      if (!c)
        throw new Error(`[ChannelServer]: Channel ${n} not found!`);
      h = c.call(r, i);
    } catch (a) {
      h = Promise.reject(a);
    }
    h.then((a) => {
      this._sendResponse({ seq: e.seq, type: 201, data: a });
    }).catch((a) => {
      a instanceof Error ? this._sendResponse({ seq: e.seq, type: 202, data: a.message }) : this._sendResponse({ seq: e.seq, type: 202, data: String(a) });
    });
  }
  _onSubscribe(e) {
    const { channelName: n, seq: r } = e, i = this._channels.get(n);
    try {
      if (!i)
        throw new Error(`[ChannelServer]: Channel ${n} not found!`);
      const h = i.subscribe(e.method, e.args).subscribe({
        next: (a) => {
          this._sendResponse({ seq: r, type: 300, data: a });
        },
        error: (a) => {
          this._sendResponse({ seq: r, type: 301, data: a.message }), this._sendResponse({
            seq: r,
            type: 302
            /* SUBSCRIBE_COMPLETE */
          });
        },
        complete: () => {
          this._sendResponse({
            seq: r,
            type: 302
            /* SUBSCRIBE_COMPLETE */
          });
        }
      });
      this._subscriptions.set(e.seq, h);
    } catch (c) {
      c instanceof Error ? this._sendResponse({ seq: e.seq, type: 301, data: c.message }) : this._sendResponse({ seq: e.seq, type: 301, data: String(c) });
    }
  }
  _onUnsubscribe(e) {
    const n = this._subscriptions.get(e.seq);
    n && (n.unsubscribe(), this._subscriptions.delete(e.seq));
  }
  _sendResponse(e) {
    this._protocol.send(e);
  }
}
const R = O("IRPCChannelService");
class B {
  constructor(t) {
    o(this, "_client");
    o(this, "_server");
    this._client = new he(t), this._server = new _e(t);
  }
  requestChannel(t) {
    return this._client.getChannel(t);
  }
  registerChannel(t, e) {
    this._server.registerChannel(t, e);
  }
}
var le = Object.defineProperty, ue = Object.getOwnPropertyDescriptor, pe = (s, t, e, n) => {
  for (var r = n > 1 ? void 0 : n ? ue(t, e) : t, i = s.length - 1, c; i >= 0; i--)
    (c = s[i]) && (r = (n ? c(t, e, r) : c(r)) || r);
  return n && r && le(t, e, r), r;
}, l = (s, t) => (e, n) => t(e, n, s);
let p = class extends q {
  constructor(t, e, n, r, i) {
    super();
    o(this, "_remoteInstanceService");
    o(this, "_syncingUnits", /* @__PURE__ */ new Set());
    o(this, "_syncingMutations", /* @__PURE__ */ new Set());
    this._injector = t, this._commandService = e, this._univerInstanceService = n, this._rpcChannelService = r, this._remoteSyncService = i, this._initRPCChannels(), this._init();
  }
  registerSyncingMutations(t) {
    this._syncingMutations.add(t.id);
  }
  _initRPCChannels() {
    this._rpcChannelService.registerChannel(j, z(this._remoteSyncService)), this._injector.add([
      S,
      { useFactory: () => G(this._rpcChannelService.requestChannel(x)) }
    ]), this._remoteInstanceService = this._injector.get(S);
  }
  _init() {
    this._univerInstanceService.getTypeOfUnitAdded$(u.UNIVER_SHEET).pipe(f(this.dispose$)).subscribe((t) => {
      this._syncingUnits.add(t.getUnitId()), this._remoteInstanceService.createInstance({
        unitID: t.getUnitId(),
        type: u.UNIVER_SHEET,
        snapshot: t.getSnapshot()
      });
    }), this._univerInstanceService.getTypeOfUnitDisposed$(u.UNIVER_SHEET).pipe(f(this.dispose$)).subscribe((t) => {
      this._syncingUnits.delete(t.getUnitId()), this._remoteInstanceService.disposeInstance({
        unitID: t.getUnitId()
      });
    }), this.disposeWithMe(this._commandService.onCommandExecuted((t, e) => {
      const { type: n, params: r, id: i } = t, c = (r == null ? void 0 : r.unitId) || "";
      n === H.MUTATION && // only sync mutations to the worker thread
      (!c || this._syncingUnits.has(c)) && // do not sync mutations from the web worker back to the web worker
      !(e != null && e.fromSync) && // do not sync mutations those are not meant to be synced
      this._syncingMutations.has(i) && this._remoteInstanceService.syncMutation({ mutationInfo: t });
    }));
  }
};
p = pe([
  W(L.Starting, p),
  l(0, g(C)),
  l(1, b),
  l(2, N),
  l(3, R),
  l(4, y)
], p);
var de = Object.defineProperty, ve = Object.getOwnPropertyDescriptor, me = (s, t, e, n) => {
  for (var r = n > 1 ? void 0 : n ? ve(t, e) : t, i = s.length - 1, c; i >= 0; i--)
    (c = s[i]) && (r = (n ? c(t, e, r) : c(r)) || r);
  return n && r && de(t, e, r), r;
}, v = (s, t) => (e, n) => t(e, n, s);
let d = class extends K {
  constructor(t, e, n, r) {
    super();
    o(this, "_remoteSyncService");
    this._injector = t, this._remoteInstanceService = e, this._commandService = n, this._rpcChannelService = r, this._initRPCChannels(), this._init();
  }
  _initRPCChannels() {
    this._rpcChannelService.registerChannel(x, z(this._remoteInstanceService)), this._injector.add([
      y,
      { useFactory: () => G(this._rpcChannelService.requestChannel(j)) }
    ]), this._remoteSyncService = this._injector.get(y);
  }
  _init() {
    this.disposeWithMe(
      // Mutations executed on the main thread should be synced to the worker thread.
      this._commandService.onCommandExecuted((t, e) => {
        t.type === H.MUTATION && !(e != null && e.fromSync) && this._remoteSyncService.syncMutation({
          mutationInfo: t
        });
      })
    );
  }
};
d = me([
  W(L.Starting, d),
  v(0, g(C)),
  v(1, S),
  v(2, b),
  v(3, R)
], d);
function fe() {
  return {
    send(s) {
      postMessage(s);
    },
    onMessage: new $((s) => {
      const t = (e) => {
        s.next(e.data);
      };
      return addEventListener("message", t), () => removeEventListener("message", t);
    }).pipe(A(1))
  };
}
function ye(s) {
  return {
    send(t) {
      s.postMessage(t);
    },
    onMessage: new $((t) => {
      const e = (n) => {
        t.next(n.data);
      };
      return s.addEventListener("message", e), () => s.removeEventListener("message", e);
    }).pipe(A(1))
  };
}
var Se = Object.defineProperty, be = Object.getOwnPropertyDescriptor, k = (s, t, e, n) => {
  for (var r = n > 1 ? void 0 : n ? be(t, e) : t, i = s.length - 1, c; i >= 0; i--)
    (c = s[i]) && (r = (n ? c(t, e, r) : c(r)) || r);
  return n && r && Se(t, e, r), r;
}, J = (s, t) => (e, n) => t(e, n, s), P;
let D = (P = class extends V {
  constructor(s, t) {
    super(), this._config = s, this._injector = t;
  }
  async onStarting(s) {
    const { workerURL: t } = this._config, e = t instanceof Worker ? t : new Worker(t), n = ye(e);
    [
      [
        R,
        {
          useFactory: () => new B(n)
        }
      ],
      [p],
      [y, { useClass: M }]
    ].forEach((i) => s.add(i)), s.get(p);
  }
}, o(P, "pluginName", "UNIVER_RPC_MAIN_THREAD_PLUGIN"), P);
D = k([
  J(1, g(C))
], D);
var E;
let T = (E = class extends V {
  constructor(s, t) {
    super(), this._config = s, this._injector = t;
  }
  onStarting(s) {
    [
      [d],
      [
        R,
        {
          useFactory: () => new B(fe())
        }
      ],
      [S, { useClass: U }]
    ].forEach((t) => s.add(t)), s.get(d);
  }
}, o(E, "pluginName", "UNIVER_RPC_WORKER_THREAD_PLUGIN"), E);
T = k([
  J(1, g(C))
], T);
export {
  he as ChannelClient,
  _e as ChannelServer,
  B as ChannelService,
  p as DataSyncPrimaryController,
  d as DataSyncReplicaController,
  R as IRPCChannelService,
  S as IRemoteInstanceService,
  y as IRemoteSyncService,
  x as RemoteInstanceServiceName,
  M as RemoteSyncPrimaryService,
  j as RemoteSyncServiceName,
  D as UniverRPCMainThreadPlugin,
  T as UniverRPCWorkerThreadPlugin,
  U as WebWorkerRemoteInstanceService
};
